universal-login-relayer/lib/services/CounterfactualIdentityService.js

```diff
@@ -0,0 +1,125 @@
+import Identity from 'universal-login-contracts/build/Identity';
+import {addressToBytes32} from '../utils/utils';
+import ethers, {utils, Wallet} from 'ethers';
+import defaultDeployOptions from '../config/defaultDeployOptions';
+
+
+class CounterfactualIdentityService {
+  constructor(wallet, ensService, authorisationService, hooks, provider) {
+    this.wallet = wallet;
+    this.abi = Identity.interface;
+    this.ensService = ensService;
+    this.authorisationService = authorisationService;
+    this.codec = new utils.AbiCoder();
+    this.hooks = hooks;
+    this.provider = provider;
+    this.minimumWeiInContractBeforeDeploy = 1;
+    this.deploymentValue = utils.bigNumberify(utils.parseEther('0.4'));
+  }
+
+  async create(managementKey, ensName, overrideOptions = {}) {
+    let deployTransaction = this._createDeployTransaction(managementKey, ensName);
+
+    deployTransaction = this._setupDeployTransaction(deployTransaction, overrideOptions);
+
+    const signedTransaction = this._signDeployTransaction(deployTransaction);
+
+    const counterfactualTransaction = this._counterfactualizeSignedTransaction(signedTransaction);
+
+    const fullCounterfactualData = await this._extractCounterfactualParams(counterfactualTransaction);
+
+    this.hooks.emit('counterfactuallyCreated', {address: fullCounterfactualData.counterfactualContractAddress});
+
+    return fullCounterfactualData;
+  }
+
+  _createDeployTransaction(managementKey, ensName) {
+    const key = addressToBytes32(managementKey);
+    const bytecode = `0x${Identity.bytecode}`;
+    const ensArgs = this.ensService.argsFor(ensName);
+    const args = [key, ...ensArgs];
+    return ethers.Contract.getDeployTransaction(bytecode, this.abi, ...args);
+  }
+
+  _setupDeployTransaction(deployTransaction, overrideOptions) {
+    const newDeployTransaction = {
+      ...deployTransaction,
+      ...defaultDeployOptions,
+      ...overrideOptions
+    };
+    return newDeployTransaction;
+  }
+
+  _signDeployTransaction(deployTransaction) {
+    return this.wallet.sign(deployTransaction);
+  }
+
+  _counterfactualizeSignedTransaction(signedTransaction) {
+    const signedTransNoRSV = signedTransaction.substring(0, signedTransaction.length - 134);
+
+    let randomS = utils.keccak256(utils.randomBytes(3));
```

Comment:
I wonder where number 3 here came from?

Reply:
Length of the random bytes generated. I will increase it.